engine/src/main/java/org/pentaho/di/trans/steps/getfilesrowscount/GetFilesRowsCount.java

```diff
@@ -0,0 +1,399 @@
+                  prevCR = true;
+                } else if ( buf[i] == '\n' ) {
+                  // we have a line feed
+                  // let's see if we had previously a carriage return
+                  if ( prevCR ) {
+                    // we have a carriage return followed by a line feed
+                    data.rownr++;
+                    // Maybe we won't have data after
+                    data.foundData = false;
+                    prevCR = false;
+                  }
+                }
+              } else {
+                // we have another char (other than \n , \r)
+                prevCR = false;
+              }
+
+            } else {
+              if ( buf[i] == data.separator ) {
+                data.rownr++;
+                // Maybe we won't have data after
+                data.foundData = false;
+              }
+            }
+          }
+        }
+      }
+      if ( isDetailed() ) {
+        logDetailed( BaseMessages.getString( PKG, "GetFilesRowsCount.Log.RowsInFile", data.file.toString(), ""
+          + data.rownr ) );
+      }
+    } catch ( Exception e ) {
+      throw new KettleException( e );
+    } finally {
+      // Close inputstream - not used except for counting
+      if ( data.fr != null ) {
+        BaseStep.closeQuietly( data.fr );
+        data.fr = null;
+      }
+    }
+
+  }
+
+  private boolean openNextFile() {
+    if ( data.last_file ) {
+      return false; // Done!
+    }
+
+    try {
+      if ( !meta.isFileField() ) {
+        if ( data.filenr >= data.files.nrOfFiles() ) {
+          // finished processing!
+
+          if ( log.isDetailed() ) {
+            logDetailed( BaseMessages.getString( PKG, "GetFilesRowsCount.Log.FinishedProcessing" ) );
+          }
+          return false;
+        }
+
+        // Is this the last file?
+        data.last_file = ( data.filenr == data.files.nrOfFiles() - 1 );
+        data.file = data.files.getFile( (int) data.filenr );
+
+      } else {
+        data.readrow = getRow(); // Get row from input rowset & set row busy!
+        if ( data.readrow == null ) {
+          if ( log.isDetailed() ) {
+            logDetailed( BaseMessages.getString( PKG, "GetFilesRowsCount.Log.FinishedProcessing" ) );
+          }
+          return false;
+        }
+
+        if ( first ) {
+          first = false;
+
+          data.inputRowMeta = getInputRowMeta();
+          data.outputRowMeta = data.inputRowMeta.clone();
+          meta.getFields( data.outputRowMeta, getStepname(), null, null, this, repository, metaStore );
+
+          // Get total previous fields
+          data.totalpreviousfields = data.inputRowMeta.size();
+
+          // Check is filename field is provided
+          if ( Utils.isEmpty( meta.setOutputFilenameField() ) ) {
+            logError( BaseMessages.getString( PKG, "GetFilesRowsCount.Log.NoField" ) );
+            throw new KettleException( BaseMessages.getString( PKG, "GetFilesRowsCount.Log.NoField" ) );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Exception [KettleException] is explicitly thrown on this line but is caught in the outer catch clause with generic exception on line [281].  Consider adding a catch clause of this specific exception in the outer try statement.